### PR TITLE
Generalize pooling accumulator inference

### PR DESCRIPTION
**Pull request by jmduarte**
Currently, pooling uses a wider data type for accumulation inferred based on the input type and the number of elements. However, we only implemented this for `ap_fixed<W,I>` and `ap_int<W>` and the template matching fails for `ap_uint<W>` and `ap_ufixed<W,I,Q,O>` and `ap_fixed<W,I,Q,O>` (if `Q` and `O` are not the default values). This PR (hopefully) makes it so it matches all those additional types.

## Type of change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Tests

Example before this fix:
profiling_global_average_pooling1d (1)

Example after this fix: 
profiling_global_average_pooling1d (2)

## Checklist

- [x] I have read the [guidelines for contributing](https://github.com/fastmachinelearning/hls4ml/blob/main/CONTRIBUTING.md).
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have installed and run `pre-commit` on the files I edited or added.
- [ ] I have added tests that prove my fix is effective or that my feature works.

@Brainz22 @ddiaz006 @AnthonyAportela @danprim

### CODE REVIEW

**Comment by jmduarte** [comment]
Do we need a corresponding fix for Quartus?

https://github.com/fastmachinelearning/hls4ml/blob/f41951c57883022362dd4ffbe6ce3ed4efe36811/hls4ml/templates/quartus/firmware/nnet_utils/nnet_pooling.h#L38-L74

File: hls4ml/templates/vivado/nnet_utils/nnet_pooling.h [hunk 1]
```diff
@@ -27,15 +27,39 @@ template <int W, int N> ap_int<W> avg(ap_int<W> (&x)[N]) {
     return tmp;
 }
 
-template <int W, int I, int N> ap_fixed<W, I> avg(ap_fixed<W, I> (&x)[N]) {
+template <int W, int N> ap_int<W> avg(ap_uint<W> (&x)[N]) {
```
Comment (vloncar) [inline hls4ml/templates/vivado/nnet_utils/nnet_pooling.h:30]: didn't you mean to return `ap_uint<W>` here?

File: hls4ml/templates/vivado/nnet_utils/nnet_pooling.h [hunk 2]
```diff
@@ -27,15 +27,39 @@ template <int W, int N> ap_int<W> avg(ap_int<W> (&x)[N]) {
...
+    for (int i = 0; i < N; i++) {
+        tmp += x[i];
+    }
+    tmp /= N;
+    // Now cast back to original type
+    ap_fixed<W, I, Q, O> y = tmp;
+    return y;
+}
+
+template <int W, int I, int N, ap_q_mode Q, ap_o_mode O> ap_ufixed<W, I, Q, O> avg(ap_ufixed<W, I, Q, O> (&x)[N]) {
```
Comment (vloncar) [inline hls4ml/templates/vivado/nnet_utils/nnet_pooling.h:54]: Technically we're still missing cases with the number of saturation bits, e.g., `ap_[u]fixed<W, I, Q, O, N>`

File: hls4ml/templates/vivado/nnet_utils/nnet_pooling.h (hunk 1 above)
Comment (jmduarte) [reply hls4ml/templates/vivado/nnet_utils/nnet_pooling.h]: Fixed

File: hls4ml/templates/vivado/nnet_utils/nnet_pooling.h (hunk 2 above)
Comment (jmduarte) [reply hls4ml/templates/vivado/nnet_utils/nnet_pooling.h]: Updated

**Comment by jmduarte** [comment]
pre-commit.ci autofix

**Comment by calad0i** [comment]
Currently, there is an `accum_t` set for pooling layers which is only effective for `io_stream` implementation: [vivado_backend.py#L376-L380](https://github.com/calad0i/hls4ml/blob/eb70a61d0da77d880de055065e3837de5846cc72/hls4ml/backends/vivado/vivado_backend.py#L376-L380), [nnet_pooling_stream.h#L386](https://github.com/calad0i/hls4ml/blob/eb70a61d0da77d880de055065e3837de5846cc72/hls4ml/templates/vivado/nnet_utils/nnet_pooling_stream.h#L386). (Which could also be problematic at the moment, see #917.)

Maybe we should converge on whether to set pooling accumulator size with `accum_t` or cpp template matching? @vloncar

**Comment by jmitrevs** [comment]
> Do we need a corresponding fix for Quartus?
> 
> https://github.com/fastmachinelearning/hls4ml/blob/f41951c57883022362dd4ffbe6ce3ed4efe36811/hls4ml/templates/quartus/firmware/nnet_utils/nnet_pooling.h#L38-L74

Actually I think it also needs to be done for vitis (which has its own implementation) in addition to quartus.

**Comment by jmitrevs** [comment]
What did we decide to do about accum_t vs template matching?

**Comment by jmitrevs** [comment]
My general feeling is to prefer the #973 style solution, though implementation details may still need to be addressed.

**Comment by jmitrevs** [comment]
We decided to go with #973 so I will close this PR.

**Pull request closed without merging**